src/python/grpcio/grpc/experimental/aio/_call.py

```diff
@@ -153,116 +152,67 @@ class Call(_base_call.Call):
     _code: grpc.StatusCode
-    _status: Awaitable[cygrpc.AioRpcStatus]
-    _initial_metadata: Awaitable[MetadataType]
-    _locally_cancelled: bool
     _cython_call: cygrpc._AioCall
-    _done_callbacks: List[DoneCallbackType]
-
-    def __init__(self, cython_call: cygrpc._AioCall) -> None:
-        self._loop = asyncio.get_event_loop()
-        self._code = None
-        self._status = self._loop.create_future()
-        self._initial_metadata = self._loop.create_future()
-        self._locally_cancelled = False
+
+    def __init__(self, cython_call: cygrpc._AioCall,
+                 loop: asyncio.AbstractEventLoop) -> None:
+        self._loop = loop
         self._cython_call = cython_call
-        self._done_callbacks = []
 
     def __del__(self) -> None:
-        if not self._status.done():
-            self._cancel(
-                cygrpc.AioRpcStatus(cygrpc.StatusCode.cancelled,
-                                    _GC_CANCELLATION_DETAILS, None, None))
+        if not self._cython_call.done():
+            self._cancel(_GC_CANCELLATION_DETAILS)
 
     def cancelled(self) -> bool:
-        return self._code == grpc.StatusCode.CANCELLED
+        return self._cython_call.cancelled()
 
-    def _cancel(self, status: cygrpc.AioRpcStatus) -> bool:
+    def _cancel(self, details: str) -> bool:
         """Forwards the application cancellation reasoning."""
-        if not self._status.done():
-            self._set_status(status)
-            self._cython_call.cancel(status)
+        if not self._cython_call.done():
+            self._cython_call.cancel(details)
             return True
         else:
             return False
 
     def cancel(self) -> bool:
-        return self._cancel(
-            cygrpc.AioRpcStatus(cygrpc.StatusCode.cancelled,
-                                _LOCAL_CANCELLATION_DETAILS, None, None))
+        return self._cancel(_LOCAL_CANCELLATION_DETAILS)
 
     def done(self) -> bool:
-        return self._status.done()
+        return self._cython_call.done()
 
     def add_done_callback(self, callback: DoneCallbackType) -> None:
-        if self.done():
-            callback(self)
-        else:
-            self._done_callbacks.append(callback)
+        cb = partial(callback, self)
+        self._cython_call.add_done_callback(cb)
 
     def time_remaining(self) -> Optional[float]:
         return self._cython_call.time_remaining()
 
     async def initial_metadata(self) -> MetadataType:
-        return await self._initial_metadata
+        return await self._cython_call.initial_metadata()
 
     async def trailing_metadata(self) -> MetadataType:
-        return (await self._status).trailing_metadata()
+        return (await self._cython_call.status()).trailing_metadata()
 
     async def code(self) -> grpc.StatusCode:
-        await self._status
-        return self._code
+        cygrpc_code = (await self._cython_call.status()).code()
+        return _common.CYGRPC_STATUS_CODE_TO_STATUS_CODE[cygrpc_code]
```

Comment:
Could we challenge the idea of having to support the `add_done_callback` by telling the user that Asyncio already provides a unified - and most likely more complete - way of scheduling callbacks by warpping a coro into a Task? For example:

```python
task = asyncio.ensure_future(call.code())
task.add_done_callback(functools.partial(your_callback, call))
```

Reply:
I didn't see much difference in functionality between these two choices. We can have a discussion about it. Is there anything I missed for scheduling callbacks?

Three reasons I would like to keep it in our layer: 1) performance, as you found out, additional future wrapper is costly; 2) consistency, since we are trying to mimic what `asyncio.Task` does. 3) less-verbose.

Reply:
My point is quite basic, not implement what is already provided by the framework, which is for sure battle-tested and in our case also implements for free the notification in case of error.

Reply:
I understand your point, and the value of neat surface interfaces. If we code the callback in snippets, the latter version looks easier to use to me. To date, the additional callback mechanism is not too complex to manage. @gnossen for additional vote.

```python
def your_callback(call, code_future): pass

task = asyncio.ensure_future(call.code())
task.add_done_callback(functools.partial(your_callback, call))
```
```python
def your_callback(call): pass

call.add_done_callback(your_callback)
```